### PR TITLE
Improve handling of test interruptions (fix #516)

### LINKED ISSUE #516: fire test_interrupt as early as possible

### PR DESCRIPTION
**Pull request by ayalash**

### CODE REVIEW

File: doc/changelog.rst [hunk 1]
```diff
@@ -1,6 +1,7 @@
 Changelog
 =========
 
+* :bug:`516 major` Fire test_interrupt earlier & setting session as interrupted as well
```
Comment (vmalloc) [inline doc/changelog.rst]: Fire test_interrupt earlier and properly mark session as interrupted when a test is interrupted

File: slash/core/result.py [hunk 2]
```diff
@@ -224,7 +241,10 @@ def __repr__(self):
 
 
 class GlobalResult(Result):
-    pass
+
+    def is_global_result(self):
+        return False
```
Comment (vmalloc) [inline slash/core/result.py]: Ummm.... I think you meant `True`

File: slash/core/result.py [hunk 3]
```diff
@@ -64,7 +68,20 @@ def add_exception(self, exc_info=None):
...
             self.add_error(exc_info=exc_info)
         else:
-            self.mark_interrupted()
+            # Assume interrupted
+            session_result = context.session.results.global_result
+            interrupted_session = session_result.is_interrupted()
+
+            if not self.is_global_result():
+                # Test was interrupted
+                interrupted_test = self.is_interrupted()
```
Comment (vmalloc) [inline slash/core/result.py]: Recommend calling the variable `was_interrupted`

File: doc/changelog.rst (hunk 1 above)
Comment (vmalloc) [inline doc/changelog.rst]: Fire test_interrupt earlier and properly mark session as interrupted when a test is interrupted

File: slash/core/result.py (hunk 2 above)
Comment (vmalloc) [inline slash/core/result.py]: Ummm... I think you meant `True`

File: slash/runner.py [hunk 4]
```diff
@@ -120,10 +119,6 @@ def _run_single_test(test, test_iterator):
 
                 except context.session.get_skip_exception_types():
                     pass
-                except INTERRUPTION_EXCEPTIONS:
```
Comment (vmalloc) [inline slash/runner.py]: Why was this deleted? I think it's still needed...

File: tests/test_interruptions.py [hunk 5]
```diff
@@ -104,6 +125,12 @@ def session_interrupt():
 
 
 @pytest.fixture
+def test_interrupt():
```
Comment (vmalloc) [inline tests/test_interruptions.py]: Please call this test_interrupt_callback or test_interrupt_hook. It is confusing along with the `test_interrupt` hook...

File: slash/runner.py (hunk 4 above)
Comment (ayalash) [inline slash/runner.py]: Why is this still needed?
All it did is firing the `test_interrupt` hook, but now we're firing it once the interruption is caught. What am I missing?

File: slash/core/result.py (hunk 2 above)
Comment (ayalash) [inline slash/core/result.py]: Oops, silly me. I fixed it (not pushed yet) & added a unittest for it. Sorry!

File: slash/runner.py (hunk 4 above)
Comment (vmalloc) [inline slash/runner.py]: @ayalash oops, my bad... Was sure the previous code didn't reraise it... ignore it.

**Comment by ayalash** [comment]
@vmalloc I pushed a new commit with the fixes. Please review again.

**Comment by vmalloc** [comment]
Thanks!
